File: LLMgeneratedCode/phi4-3.py
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt

# Load feature data and metadata
df_features_train = pd.read_csv('df_training.csv', index_col=0)
df_metadata_train = pd.read_csv('ano_training.csv')

df_features_test = pd.read_csv('df_test.csv', index_col=0)
df_metadata_test = pd.read_csv('ano_test.csv')

# Merge feature data with metadata
merged_train = df_features_train.merge(df_metadata_train, left_index=True, right_on='Sample_ID')
merged_test = df_features_test.merge(df_metadata_test, left_index=True, right_on='Sample_ID')

# Prepare the training and test datasets
X_train = merged_train.drop(columns=['Sample_ID', 'gestational_age'])
y_train = merged_train['gestational_age']

X_test = merged_test.drop(columns=['Sample_ID', 'gestational_age'])
y_test = merged_test['gestational_age']

# Fit a predictive model (Linear Regression in this case)
model = LinearRegression()
model.fit(X_train, y_train)

# Predict on the test set
y_pred = model.predict(X_test)

# Calculate RMSE
rmse = mean_squared_error(y_test, y_pred, squared=False)
print(f'Root Mean Squared Error (RMSE): {rmse}')

# Generate scatter plot of predicted vs actual gestational age values
plt.figure(figsize=(10, 6))
plt.scatter(y_test, y_pred, alpha=0.5)
plt.xlabel('Actual Gestational Age')
plt.ylabel('Predicted Gestational Age')
plt.title('Predicted vs Actual Gestational Ages')
plt.plot([y.min(), y.max()], [y.min(), y.max()], 'k--', lw=2)  # Diagonal line
plt.text(0.1, 0.9 * max(y_test.max(), y_pred.max()), f'RMSE: {rmse:.2f}', transform=plt.gca().transAxes)
plt.grid(True)
plt.savefig("./out2/phi4-3.py.pdf", format="pdf")

File: LLMgeneratedCode/phi4-3.R
# Load necessary libraries
library(tidyverse)
library(caret)

# Load the data
df_training <- read.csv("df_training.csv", stringsAsFactors = FALSE)
ano_training <- read.csv("ano_training.csv", stringsAsFactors = FALSE)

df_test <- read.csv("df_test.csv", stringsAsFactors = FALSE)
ano_test <- read.csv("ano_test.csv", stringsAsFactors = FALSE)

# Merge the feature data with metadata using Sample_ID
training_data <- df_training %>%
  column_to_rownames(var = "Feature") %>%
  as_tibble() %>%
  pivot_longer(-c(name), names_to = "Sample_ID", values_to = "Methylation") %>%
  left_join(ano_training, by = "Sample_ID")

test_data <- df_test %>%
  column_to_rownames(var = "Feature") %>%
  as_tibble() %>%
  pivot_longer(-c(name), names_to = "Sample_ID", values_to = "Methylation") %>%
  left_join(ano_test, by = "Sample_ID")

# Prepare the data for modeling
train_df <- training_data %>%
  spread(name, Methylation) %>%
  select(-Sample_ID)

test_df <- test_data %>%
  spread(name, Methylation) %>%
  select(-Sample_ID)

# Extract gestational age as response variable
y_train <- training_data$GA
y_test <- test_data$GA

# Fit a linear regression model
model <- lm(GA ~ ., data = train_df)

# Make predictions on the test set
predictions <- predict(model, newdata = test_df)

# Calculate RMSE
rmse_value <- sqrt(mean((predictions - y_test)^2))
cat("Root Mean Squared Error (RMSE) on Test Set:", rmse_value, "\n")

# Generate scatter plot of predicted vs actual gestational age values
plot(y_test, predictions,
     xlab = "Actual Gestational Age",
     ylab = "Predicted Gestational Age",
     main = "Predicted vs Actual Gestational Age",
     pch = 19)

abline(0, 1, col = "red", lty = 2) # Line of equality

legend("topleft", legend = paste("RMSE:", round(rmse_value, 4)), bty = "n")
